### SENTIMENT ANALYSIS

In [543]:
import requests
from bs4 import BeautifulSoup as bs
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 1. Choose a particular product in the cellular phone category - 

   
   ##                              ------------       iphone13     --------------

In [ ]:
#for i in range(1,1001):
    
    url = 'https://www.flipkart.com/apple-iphone-13-green-128-gb/product-reviews/itm18a55937b2607?pid=MOBGC9VGSU9DWGJZ&lid=LSTMOBGC9VGSU9DWGJZTOZYKQ&marketplace=FLIPKART'+'&page='+str(i)
        
    review_html = requests.get(url)
    
    soup = bs(review_html.content, 'html.parser')
    
    for data in soup.findAll('div',class_='_27M-vq'):

        rating=data.find('div', attrs={'class':'_3LWZlK'})
        review=data.find('div', attrs={'class':'t-ZTKy'})
        page.append(i)

        if "...READ MORE" not in review.text:
            all_ratings.append(rating.text)
            all_reviews.append((review.text).replace("READ MORE",""))
        else:
            continue
            
ratings=[]
for i in all_ratings:
    ratings.append(int(i))       
df=pd.DataFrame(list(zip(page,all_reviews,ratings)),columns =['Page','Reviews', 'Ratings'])
df.to_csv("iphone13_fipkart.csv",index=False)

## 2. Scrape product reviews from Flipkart

In [45]:
df=pd.read_csv("iphone13_fipkart.csv")
df

Page                                            Reviews  Ratings
0        1  Awesome battery backup amazing camera, a all d...        5
1        1  Just upgraded from 11 to iPhone 13 ..cinematic...        5
2        1  Picture quality is amazing you do not need to ...        5
3        1  Awesome camera Amazing battery life Super xdr ...        5
4        1  Performance very good Display very good Batter...        5
...    ...                                                ...      ...
4218   944                                  Third class phone        1
4219   944  Upgraded from X to XIII. Amazing camera. Built...        5
4220   944                             Excellent item to buy.        5
4221   945              Phone is fast , classy and good phone        5
4222   945             Apple never disappoints. Amazing phone        5

[4223 rows x 3 columns]

In [46]:
df.Ratings.value_counts()

5    3480
4     464
1     132
3     115
2      32
Name: Ratings, dtype: int64

## 3. Read a few reviews from each category (1 star, 2 stars, 3 stars, 4 stars and 5 stars)

In [47]:
rating_1=df[df['Ratings']==1].sample(100,random_state=10)
rating_2=df[df['Ratings']==2].sample(32,random_state=10)
rating_3=df[df['Ratings']==3].sample(115,random_state=10)
rating_4=df[df['Ratings']==4].sample(200,random_state=10)
rating_5=df[df['Ratings']==5].sample(300,random_state=10)

In [48]:
rating_5.tail(2)

Page                                        Reviews  Ratings
723    130                                           good        5
1694   425  The best and thank u so much for the offer 🎉😍        5

In [49]:
df_reviews=pd.DataFrame()
df_reviews=df_reviews.append([rating_1,rating_2,rating_3,rating_4,rating_5])
df_reviews.reset_index(inplace=True,drop=True)
df_reviews

Page                                            Reviews  Ratings
0     799  Facing different issues since after the very f...        1
1     684                                 Sim is not working        1
2     176       My iphone 13 is look Apple I'd is not change        1
3     918      Cable doesn't work now there's no replacement        1
4     853                            Product heating problem        1
..    ...                                                ...      ...
742   603                          Amazing Love the camera 📸        5
743    41                                               Nice        5
744   521                                          Brilliant        5
745   130                                               good        5
746   425      The best and thank u so much for the offer 🎉😍        5

[747 rows x 3 columns]

In [50]:
len(df_reviews.Reviews)

747

## 4. Create a text preprocessing pipeline for review cleaning (DO NOT remove stopwords)

In [51]:
import string
p = string.punctuation
p

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [52]:
p='"#$%&\'()*+-/:;<=>?@[\\]^_`{|}~'

In [53]:
#1. Converting the text into lower cases
for i in range(len(df_reviews.Reviews)):
    df_reviews.Reviews[i] = df_reviews.Reviews[i].lower()
#---------------------------------------------------------------------------------------------------

#2. Remove Punctuations/special symbols and 
#import string
#p = string.punctuation
remv_punc = str.maketrans("", "", p)

#.Removing punctuation from all the reviewa
for i in range(len(df_reviews.Reviews)):
    df_reviews.Reviews[i] = df_reviews.Reviews[i].translate(remv_punc)
    
#---------------------------------------------------------------------------------------------------

#3. Remove white spaces
for i in range(len(df_reviews.Reviews)):
    df_reviews.Reviews[i] = df_reviews.Reviews[i].replace("  ", " ").strip()
    
     # Removing numbers
    
    df_reviews.Reviews[i] = ''.join([w for w in df_reviews.Reviews[i] if not w.isdigit()])
    
    # Removing weblinks 
    #import re
    #re.sub(r'''(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'".,<>?«»“”‘’]))''', " ",df_reviews.Reviews[i])

    
#---------------------------------------------------------------------------------------------------

#4. Stemming
#from nltk.stem import PorterStemmer
#ps = PorterStemmer()
#for i in range(len(df_reviews.Reviews)):
#    df_reviews.Reviews[i] = " ".join([ps.stem(w) for w in df_reviews.Reviews[i].split()])

df_reviews.head()

Page                                            Reviews  Ratings
0   799  facing different issues since after the very f...        1
1   684                                 sim is not working        1
2   176          my iphone  is look apple id is not change        1
3   918        cable doesnt work now theres no replacement        1
4   853                            product heating problem        1

In [54]:
# Removing emojis
#for i in range(len(df_reviews.Reviews)):
#    df_reviews.Reviews[i] = df_reviews.Reviews[i].encode('ascii', 'ignore').decode('ascii')

In [55]:
df_reviews

Page                                            Reviews  Ratings
0     799  facing different issues since after the very f...        1
1     684                                 sim is not working        1
2     176          my iphone  is look apple id is not change        1
3     918        cable doesnt work now theres no replacement        1
4     853                            product heating problem        1
..    ...                                                ...      ...
742   603                          amazing love the camera 📸        5
743    41                                               nice        5
744   521                                          brilliant        5
745   130                                               good        5
746   425      the best and thank u so much for the offer 🎉😍        5

[747 rows x 3 columns]

### Sentiment Analysis using VADER

In [56]:
# Getting the sentiment scores for five reviews
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

i=0
for review in df_reviews.Reviews:
    senti=SentimentIntensityAnalyzer() ### Sentiment Analysis using VADER
    print(review)
    print(senti.polarity_scores(review),"\n")
    i+=1
    if i==6:
        break

facing different issues since after the very first week itself. face lock not working, app crashes
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0} 

sim is not working
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0} 

my iphone  is look apple id is not change
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0} 

cable doesnt work now theres no replacement
{'neg': 0.268, 'neu': 0.732, 'pos': 0.0, 'compound': -0.296} 

product heating problem
{'neg': 0.574, 'neu': 0.426, 'pos': 0.0, 'compound': -0.4019} 

with in a month the phone speaker is giving some annoying noise, really disappointed with the flipkart product
{'neg': 0.259, 'neu': 0.638, 'pos': 0.102, 'compound': -0.5709} 



### Feature-Based Opinion Mining

## 5. Once the cleaning is done, use POS tagger to tag the parts of speech

#### 1. Feature Extraction
#### A. Extraction of Frequent Nouns/Noun Phrase

In [57]:
import nltk
#nltk.download("averaged_perceptron_tagger")
from nltk import word_tokenize
from nltk import pos_tag


In [58]:
#Word tokenization

tokens=[]
for review in df_reviews.Reviews:
    token=word_tokenize(review)
    tokens.append(token)

In [59]:
tokens

[['facing',
  'different',
  'issues',
  'since',
  'after',
  'the',
  'very',
  'first',
  'week',
  'itself',
  '.',
  'face',
  'lock',
  'not',
  'working',
  ',',
  'app',
  'crashes'],
 ['sim', 'is', 'not', 'working'],
 ['my', 'iphone', 'is', 'look', 'apple', 'id', 'is', 'not', 'change'],
 ['cable', 'doesnt', 'work', 'now', 'theres', 'no', 'replacement'],
 ['product', 'heating', 'problem'],
 ['with',
  'in',
  'a',
  'month',
  'the',
  'phone',
  'speaker',
  'is',
  'giving',
  'some',
  'annoying',
  'noise',
  ',',
  'really',
  'disappointed',
  'with',
  'the',
  'flipkart',
  'product'],
 ['wrost',
  'customer',
  'service',
  ',',
  'unexpected',
  'response',
  'from',
  'flipkart',
  '.',
  'fully',
  'regretted',
  ',',
  'i',
  'will',
  'uninstall',
  'the',
  'application',
  'and',
  'dont',
  'want',
  'to',
  'use',
  'flipkart',
  '.',
  'waste',
  'of',
  'time',
  ',',
  'there',
  'is',
  'no',
  'humanity',
  ',',
  'entire',
  'flipkart',
  'support',
  'i

In [60]:
#Extracting POS for all the reviews
#POS tagging

pos_tagged=[]
for i in range(len(tokens)):
    print(nltk.pos_tag(tokens[i]),"\n")
    pos_tagged.append(nltk.pos_tag(tokens[i]))
pos_tagged

[('facing', 'VBG'), ('different', 'JJ'), ('issues', 'NNS'), ('since', 'IN'), ('after', 'IN'), ('the', 'DT'), ('very', 'RB'), ('first', 'JJ'), ('week', 'NN'), ('itself', 'PRP'), ('.', '.'), ('face', 'NN'), ('lock', 'NN'), ('not', 'RB'), ('working', 'VBG'), (',', ','), ('app', 'JJ'), ('crashes', 'NNS')] 

[('sim', 'NN'), ('is', 'VBZ'), ('not', 'RB'), ('working', 'VBG')] 

[('my', 'PRP$'), ('iphone', 'NN'), ('is', 'VBZ'), ('look', 'JJ'), ('apple', 'NN'), ('id', 'NN'), ('is', 'VBZ'), ('not', 'RB'), ('change', 'NN')] 

[('cable', 'NN'), ('doesnt', 'NN'), ('work', 'NN'), ('now', 'RB'), ('theres', 'VBZ'), ('no', 'DT'), ('replacement', 'NN')] 

[('product', 'NN'), ('heating', 'NN'), ('problem', 'NN')] 

[('with', 'IN'), ('in', 'IN'), ('a', 'DT'), ('month', 'NN'), ('the', 'DT'), ('phone', 'NN'), ('speaker', 'NN'), ('is', 'VBZ'), ('giving', 'VBG'), ('some', 'DT'), ('annoying', 'VBG'), ('noise', 'NN'), (',', ','), ('really', 'RB'), ('disappointed', 'JJ'), ('with', 'IN'), ('the', 'DT'), ('flipkart

[('beware', 'NN'), ('of', 'IN'), ('hype', 'NN'), (',', ','), ('its', 'PRP$'), ('a', 'DT'), ('phone', 'NN'), ('that', 'WDT'), ('just', 'RB'), ('has', 'VBZ'), ('call', 'JJ'), ('logs', 'NNS'), ('for', 'IN'), ('days', 'NNS'), (',', ','), ('it', 'PRP'), ('does', 'VBZ'), ('not', 'RB'), ('support', 'VB'), ('truecaller', 'NN'), ('.', '.'), ('its', 'PRP$'), ('not', 'RB'), ('a', 'DT'), ('smartphone', 'NN'), ('but', 'CC'), ('a', 'DT'), ('camera', 'NN'), ('attached', 'VBN'), ('to', 'TO'), ('a', 'DT'), ('simple', 'JJ'), ('dialup', 'NN'), ('phone', 'NN'), ('.', '.')] 

[('the', 'DT'), ('screen', 'NN'), ('is', 'VBZ'), ('going', 'VBG'), ('blank', 'JJ'), (',', ','), ('not', 'RB'), ('able', 'JJ'), ('to', 'TO'), ('connect', 'VB'), ('with', 'IN'), ('call', 'NN'), ('centre', 'NN'), ('.', '.')] 

[('display', 'NN'), ('light', 'JJ'), ('destroyed', 'VBN'), ('in', 'IN'), ('days', 'NNS')] 

[('display', 'VB'), ('very', 'RB'), ('slow', 'JJ'), ('scrolling', 'NN')] 

[('front', 'JJ'), ('camera', 'NN'), ('disappoin

[('rd', 'JJ'), ('class', 'NN'), ('phone', 'NN'), ('and', 'CC'), ('service', 'NN')] 

[('from', 'IN'), ('the', 'DT'), ('first', 'JJ'), ('day', 'NN'), ('..', 'NNP'), ('device', 'NN'), ('heating', 'VBG'), ('like', 'IN'), ('iron', 'NN'), ('box', 'NN')] 

[('the', 'DT'), ('pictures', 'NNS'), ('from', 'IN'), ('back', 'JJ'), ('cam', 'NN'), ('has', 'VBZ'), ('light', 'VBN'), ('glare', 'NN'), ('.', '.'), ('this', 'DT'), ('is', 'VBZ'), ('from', 'IN'), ('the', 'DT'), ('very', 'RB'), ('beginning', 'VBG')] 

[('after', 'IN'), ('using', 'VBG'), ('minutes', 'NNS'), ('mobile', 'RB'), ('will', 'MD'), ('get', 'VB'), ('heat', 'NN')] 

[('not', 'RB'), ('as', 'IN'), ('expected', 'VBN'), ('.', '.')] 

[('not', 'RB'), ('happy', 'JJ')] 

[('it', 'PRP'), ('gets', 'VBZ'), ('heat', 'VBD'), ('up', 'RP'), ('while', 'IN'), ('charging', 'VBG'), ('plugged', 'VBN'), ('in', 'IN')] 

[('dont', 'JJ'), ('buy', 'VB'), ('this', 'DT'), ('product', 'NN'), ('my', 'PRP$'), ('mobile', 'JJ'), ('got', 'VBD'), ('damaged', 'VBN'), ('

[('after', 'IN'), ('one', 'CD'), ('month', 'NN'), ('use', 'NN'), ('limitation', 'NN'), ('of', 'IN'), ('apps', 'NN'), ('.', '.'), ('some', 'DT'), ('apps', 'NNS'), ('are', 'VBP'), ('not', 'RB'), ('available', 'JJ'), ('on', 'IN'), ('iphone', 'NN'), (',', ','), ('some', 'DT'), ('apps', 'NNS'), ('are', 'VBP'), ('paid', 'VBN'), ('for', 'IN'), ('iphone', 'NN'), ('which', 'WDT'), ('are', 'VBP'), ('free', 'JJ'), ('on', 'IN'), ('android', 'NN'), ('.', '.'), ('not', 'RB'), ('user', 'JJ'), ('friendly', 'RB'), ('.', '.'), ('no', 'DT'), ('call', 'NN'), ('recording', 'VBG'), ('function', 'NN'), ('.', '.'), ('no', 'DT'), ('option', 'NN'), ('of', 'IN'), ('going', 'VBG'), ('back', 'RB'), ('except', 'IN'), ('home', 'NN'), ('screen', 'NN'), ('.', '.'), ('camera', 'NN'), ('is', 'VBZ'), ('good', 'JJ'), ('.', '.'), ('battery', 'NN'), ('lasts', 'VBZ'), ('for', 'IN'), ('days', 'NNS'), ('.', '.'), ('siri', 'JJ'), ('option', 'NN'), ('is', 'VBZ'), ('very', 'RB'), ('useful', 'JJ'), ('.', '.'), ('flipkart', 'JJ'), 

[('open', 'JJ'), ('box', 'NN'), ('delivery', 'NN'), ('is', 'VBZ'), ('really', 'RB'), ('useful', 'JJ'), ('and', 'CC'), ('i', 'NNS'), ('have', 'VBP'), ('checked', 'VBN'), ('serial', 'JJ'), ('number', 'NN'), ('on', 'IN'), ('apple', 'NN'), ('website', 'NN'), ('to', 'TO'), ('confirm', 'VB'), ('that', 'IN'), ('i', 'JJ'), ('got', 'VBD'), ('original', 'JJ'), ('product', 'NN'), ('so', 'RB'), ('just', 'RB'), ('go', 'VB'), ('for', 'IN'), ('it', 'PRP'), ('and', 'CC'), ('i', 'VBZ'), ('got', 'VBD'), ('it', 'PRP'), ('for', 'IN'), ('k', 'NN'), ('during', 'IN'), ('bbd😁', 'NN'), ('....', 'NN'), ('i', 'NN'), ('will', 'MD'), ('update', 'VB'), ('review', 'NN'), ('if', 'IN'), ('i', 'JJ'), ('find', 'VBP'), ('any', 'DT'), ('issues', 'NNS'), ('...', ':'), ('update', 'RB'), ('after', 'IN'), ('months', 'NNS'), ('of', 'IN'), ('usage', 'JJ'), ('performance', 'NN'), ('is', 'VBZ'), ('somewhat', 'RB'), ('decreased', 'JJ'), ('and', 'CC'), ('face', 'JJ'), ('unlock', 'NN'), ('became', 'VBD'), ('slow', 'JJ'), ('so', 'RB'

[('good', 'JJ'), ('nd', 'NN'), ('see', 'VBP'), ('later', 'RB'), ('to', 'TO'), ('use', 'VB')] 

[('good', 'JJ'), ('product', 'NN')] 

[('nice', 'JJ'), ('product', 'NN')] 

[('over', 'IN'), ('all', 'DT'), ('best', 'JJS')] 

[('satisfactory', 'JJ'), ('product', 'NN'), ('for', 'IN'), ('k.thanks', 'NNS'), ('to', 'TO'), ('flipkart', 'VB'), ('.', '.')] 

[('writing', 'VBG'), ('this', 'DT'), ('review', 'NN'), ('after', 'IN'), ('using', 'VBG'), ('two', 'CD'), ('months.camera', 'NN'), ('does', 'VBZ'), ('the', 'DT'), ('job', 'NN'), ('can', 'MD'), ('rate', 'NN'), ('.', '.'), ('out', 'IN'), ('of', 'IN'), ('and', 'CC'), ('about', 'IN'), ('the', 'DT'), ('battery', 'NN'), ('iam', 'NN'), ('a', 'DT'), ('moderate', 'JJ'), ('user', 'NN'), ('can', 'MD'), ('be', 'VB'), ('used', 'VBN'), ('a', 'DT'), ('full', 'JJ'), ('day', 'NN'), (',', ','), ('but', 'CC'), ('if', 'IN'), ('you', 'PRP'), ('are', 'VBP'), ('a', 'DT'), ('heavy', 'JJ'), ('user', 'NN'), ('need', 'NN'), ('to', 'TO'), ('charge', 'VB'), ('twice', 'RB'

[('my', 'PRP$'), ('first', 'JJ'), ('impression', 'NN'), (',', ','), ('the', 'DT'), ('camera', 'NN'), ('is', 'VBZ'), ('not', 'RB'), ('that', 'RB'), ('good', 'JJ'), ('comparing', 'NN'), ('to', 'TO'), ('the', 'DT'), ('android', 'NN'), ('phone', 'NN'), ('in', 'IN'), ('this', 'DT'), ('price', 'NN'), ('range', 'NN'), ('.', '.'), ('phone', 'NN'), ('is', 'VBZ'), ('very', 'RB'), ('small', 'JJ'), ('and', 'CC'), ('cute', 'JJ'), ('but', 'CC'), ('if', 'IN'), ('you', 'PRP'), ('thought', 'VBD'), ('to', 'TO'), ('play', 'VB'), ('games', 'NNS'), ('in', 'IN'), ('this', 'DT'), ('phone', 'NN'), ('please', 'NN'), ('think', 'VB'), ('twice', 'RB'), ('before', 'IN'), ('you', 'PRP'), ('buy.', 'VBP'), ('.', '.'), ('inch', 'NN'), ('screen', 'NN'), ('is', 'VBZ'), ('not', 'RB'), ('for', 'IN'), ('the', 'DT'), ('gamers', 'NNS'), ('.', '.'), ('need', 'NN'), ('to', 'TO'), ('check', 'VB'), ('other', 'JJ'), ('features', 'NNS'), ('in', 'IN'), ('the', 'DT'), ('future', 'NN'), ('and', 'CC'), ('update', 'VB'), ('the', 'DT'),

[('good', 'JJ'), ('product', 'NN')] 

[('just', 'RB'), ('love', 'VB'), ('it', 'PRP')] 

[('ot', 'NN'), ('was', 'VBD'), ('amazing', 'VBG'), ('experience', 'NN'), ('till', 'NN'), ('now', 'RB')] 

[('osm', 'NN'), ('just', 'RB'), ('go', 'VB'), ('for', 'IN'), ('it', 'PRP')] 

[('go', 'VB'), ('for', 'IN'), ('it', 'PRP')] 

[('ok', 'JJ'), ('ok', 'NN')] 

[('no', 'DT'), ('issue', 'NN')] 

[('everything', 'NN'), ('is', 'VBZ'), ('just', 'RB'), ('awesome', 'JJ'), ('except', 'IN'), ('not', 'RB'), ('a', 'DT'), ('day', 'NN'), ('charge', 'NN')] 

[('good', 'JJ'), ('product', 'NN')] 

[('the', 'DT'), ('picture', 'NN'), ('quality', 'NN'), ('of', 'IN'), ('this', 'DT'), ('phone', 'NN'), ('is', 'VBZ'), ('so', 'RB'), ('good', 'JJ'), ('and', 'CC'), ('look', 'VB'), ('wise', 'NN'), ('totally', 'RB'), ('🔥🔥', 'JJ')] 

[('just', 'RB'), ('switch', 'VB'), ('to', 'TO'), ('iphone', 'VB'), ('from', 'IN'), ('android', 'NN'), ('.', '.'), ('very', 'RB'), ('satisfying', 'JJ'), ('performance', 'NN'), ('display', 'NN'), ('

[('nice', 'JJ')] 

[('good', 'JJ')] 

[('purchased', 'VBN'), ('from', 'IN'), ('bbd', 'NN'), ('at', 'IN'), ('k', 'NN'), (',', ','), ('awesome', 'JJ'), ('deal', 'NN'), ('awesome', 'JJ'), ('phone', 'NN'), (',', ','), ('however', 'RB'), ('slow', 'JJ'), ('charging', 'VBG'), ('phone', 'NN'), (',', ','), ('and', 'CC'), ('camera', 'VB'), ('up', 'RB'), ('to', 'TO'), ('the', 'DT'), ('mark', 'NN'), ('.', '.')] 

[('nice', 'JJ'), ('one', 'CD')] 

[('great', 'JJ')] 

[('best', 'JJS'), ('phone', 'NN'), (',', ','), ('battery', 'NN'), ('backup', 'NN'), ('is', 'VBZ'), ('also', 'RB'), ('improved', 'VBN'), ('in', 'IN'), ('this', 'DT'), ('seriesoverall', 'JJ'), ('good', 'JJ'), ('experience', 'NN')] 

[('wow', 'JJ'), ('very', 'RB'), ('nice', 'JJ'), ('phone', 'NN'), ('.', '.'), ('performance', 'NN'), (',', ','), ('battery', 'NN'), (',', ','), ('camera', 'NN'), (',', ','), ('display', 'VB'), ('totally', 'RB'), ('everything', 'NN'), ('is', 'VBZ'), ('awesome', 'JJ'), ('.', '.'), ('must', 'MD'), ('buy', 'VB'), 

[('great', 'JJ'), ('phone', 'NN')] 

[('awesome', 'JJ'), ('😎', 'NN')] 

[('really', 'RB'), ('good', 'JJ'), ('😊', 'NN'), ('fully', 'RB'), ('enjoyed', 'VBN'), ('!', '.'), ('!', '.'), ('👌😎', 'NN')] 

[('no', 'DT'), ('need', 'NN'), ('to', 'TO'), ('check', 'VB'), ('any', 'DT'), ('reviews', 'NNS'), ('.', '.'), ('just', 'RB'), ('buy', 'VB'), ('blindly', 'RB')] 

[('loved', 'VBD'), ('it', 'PRP'), ('!', '.'), ('!', '.')] 

[('awsm', 'NN')] 

[('really', 'RB'), ('loved', 'VBD'), ('it', 'PRP'), ('❣️', 'VB')] 

[('really', 'RB'), ('liked', 'VBD'), ('it', 'PRP')] 

[('great', 'JJ'), ('deal', 'NN'), ('😍😍😍thank', 'NN'), ('you', 'PRP'), ('flipkart', 'VBP'), ('🥰🥰🥰🥰', 'VB')] 

[('awesome', 'JJ'), ('experience', 'NN'), ('.', '.'), ('camera', 'NN'), ('quality', 'NN'), ('is', 'VBZ'), ('excellent', 'JJ')] 

[('the', 'DT'), ('experience', 'NN'), ('of', 'IN'), ('shifting', 'VBG'), ('to', 'TO'), ('android', 'VB'), ('theke', 'NN'), ('iphone', 'NN'), ('is', 'VBZ'), ('so', 'RB'), ('completely', 'RB'), ('different

[('just', 'RB'), ('wowww', 'NN'), ('...', ':'), ('😍', 'NN')] 

[('good', 'JJ')] 

[('just', 'RB'), ('wow', 'NN')] 

[('excellent', 'NN')] 

[('prosbrighter', 'NN'), ('displaygreat', 'NN'), ('camerascompelling', 'VBG'), ('cinematic', 'JJ'), ('video', 'NN'), ('modeclassleading', 'VBG'), ('performancevery', 'RB'), ('good', 'JJ'), ('battery', 'NN'), ('lifeconslacks', 'NNS'), ('hz', 'VBP'), ('displayslower', 'JJR'), ('charging', 'VBG'), ('than', 'IN'), ('rivalscinematic', 'JJ'), ('mode', 'NN'), ('capped', 'VBD'), ('at', 'IN'), ('p', 'NN')] 

[('best', 'JJS'), ('iphone', 'NN')] 

[('awesome', 'JJ'), ('performance', 'NN'), ('with', 'IN'), ('awesome', 'JJ'), ('phone', 'NN')] 

[('its', 'PRP$'), ('too', 'RB'), ('smooth', 'JJ'), ('and', 'CC'), ('good', 'JJ'), ('looking', 'VBG'), ('phone', 'NN'), ('with', 'IN'), ('good', 'JJ'), ('performance', 'NN'), ('a', 'DT'), ('iam', 'JJ'), ('very', 'RB'), ('happy', 'JJ'), ('tq', 'NN')] 

[('good', 'JJ'), ('phone', 'NN'), ('just', 'RB'), ('go', 'VB'), ('for',

[('excellent', 'NN')] 

[('best', 'JJS'), ('phone', 'NN'), ('that', 'WDT'), ('i', 'VBZ'), ('ever', 'RB'), ('had', 'VBN'), ('.', '.'), ('shifting', 'VBG'), ('from', 'IN'), ('android', 'JJ'), ('to', 'TO'), ('ios', 'VB'), (',', ','), ('i', 'NNS'), ('have', 'VBP'), ('a', 'DT'), ('worthy', 'JJ'), ('journey', 'NN'), ('.', '.'), ('great', 'JJ'), ('camera', 'NN'), (',', ','), ('great', 'JJ'), ('performance', 'NN'), (',', ','), ('so', 'RB'), ('easy', 'JJ'), ('to', 'TO'), ('use', 'VB'), ('.', '.'), ('recomending', 'VBG'), ('everyone', 'NN'), ('.', '.')] 

[('osm', 'NN'), ('phone', 'NN')] 

[('awesome', 'NN')] 

[('good', 'JJ')] 

[('best', 'JJS'), ('buy', 'NN'), ('of', 'IN'), ('the', 'DT'), ('year', 'NN'), (',', ','), ('out', 'IN'), ('of', 'IN'), ('the', 'DT'), ('line', 'NN'), ('battery', 'NN'), ('life', 'NN'), ('.', '.'), ('amazing', 'JJ'), ('display', 'NN'), ('and', 'CC'), ('great', 'JJ'), ('camera', 'NN'), ('.', '.')] 

[('great', 'JJ'), ('job', 'NN'), ('flipkart', 'NN'), ('.', '.'), ('thanks

[('great', 'JJ'), ('job', 'NN')] 

[('super', 'NN'), ('❤️iphone', 'NN')] 

[('unbelievable', 'JJ'), ('the', 'DT'), ('performance', 'NN'), ('really', 'RB'), ('fantastic', 'JJ'), (',', ','), ('im', 'JJ'), ('fully', 'RB'), ('enjoying', 'VBG'), ('with', 'IN'), ('it', 'PRP'), ('.', '.'), ('battery', 'NN'), ('back', 'RB'), ('up', 'IN'), ('mind', 'NN'), ('blowing', 'NN'), ('...', ':'), ('overall', 'JJ'), ('osm', 'NN')] 

[('nice', 'JJ'), ('product', 'NN')] 

[('good', 'JJ')] 

[('nice', 'JJ'), ('phone', 'NN'), ('kharid', 'NN'), ('lo', 'NN')] 

[('royal', 'NN'), ('feeling', 'NN')] 

[('awesome', 'NN')] 

[('just', 'RB'), ('wao', 'WRB'), ('😍', 'JJ'), ('amazing', 'JJ'), ('picture', 'NN'), ('quality', 'NN')] 

[('thank', 'JJ'), ('flipkart', 'NN'), ('for', 'IN'), ('very', 'RB'), ('best', 'JJS'), ('price', 'NN'), ('and', 'CC'), ('very', 'RB'), ('fast', 'RB'), ('delivery', 'NN'), ('.', '.'), ('i', 'VB'), ('phone', 'NN'), ('camera', 'NN'), ('quality', 'NN'), ('is', 'VBZ'), ('good', 'JJ'), ('.', '.')]

[[('facing', 'VBG'),
  ('different', 'JJ'),
  ('issues', 'NNS'),
  ('since', 'IN'),
  ('after', 'IN'),
  ('the', 'DT'),
  ('very', 'RB'),
  ('first', 'JJ'),
  ('week', 'NN'),
  ('itself', 'PRP'),
  ('.', '.'),
  ('face', 'NN'),
  ('lock', 'NN'),
  ('not', 'RB'),
  ('working', 'VBG'),
  (',', ','),
  ('app', 'JJ'),
  ('crashes', 'NNS')],
 [('sim', 'NN'), ('is', 'VBZ'), ('not', 'RB'), ('working', 'VBG')],
 [('my', 'PRP$'),
  ('iphone', 'NN'),
  ('is', 'VBZ'),
  ('look', 'JJ'),
  ('apple', 'NN'),
  ('id', 'NN'),
  ('is', 'VBZ'),
  ('not', 'RB'),
  ('change', 'NN')],
 [('cable', 'NN'),
  ('doesnt', 'NN'),
  ('work', 'NN'),
  ('now', 'RB'),
  ('theres', 'VBZ'),
  ('no', 'DT'),
  ('replacement', 'NN')],
 [('product', 'NN'), ('heating', 'NN'), ('problem', 'NN')],
 [('with', 'IN'),
  ('in', 'IN'),
  ('a', 'DT'),
  ('month', 'NN'),
  ('the', 'DT'),
  ('phone', 'NN'),
  ('speaker', 'NN'),
  ('is', 'VBZ'),
  ('giving', 'VBG'),
  ('some', 'DT'),
  ('annoying', 'VBG'),
  ('noise', 'NN'),
  (',', ',

### We'll tokenize, POS tag and extract nouns at once

In [161]:
#Word tokenization
from nltk import word_tokenize
token_list = []
for review in df_reviews.Reviews:
    token_list.append(word_tokenize(review.lower()))

#POS tagging
#Save the NOUNS only
#Extracting the nouns for all the reviews
#List of unique nouns
token_list

[['facing',
  'different',
  'issues',
  'since',
  'after',
  'the',
  'very',
  'first',
  'week',
  'itself',
  '.',
  'face',
  'lock',
  'not',
  'working',
  ',',
  'app',
  'crashes'],
 ['sim', 'is', 'not', 'working'],
 ['my', 'iphone', 'is', 'look', 'apple', 'id', 'is', 'not', 'change'],
 ['cable', 'doesnt', 'work', 'now', 'theres', 'no', 'replacement'],
 ['product', 'heating', 'problem'],
 ['with',
  'in',
  'a',
  'month',
  'the',
  'phone',
  'speaker',
  'is',
  'giving',
  'some',
  'annoying',
  'noise',
  ',',
  'really',
  'disappointed',
  'with',
  'the',
  'flipkart',
  'product'],
 ['wrost',
  'customer',
  'service',
  ',',
  'unexpected',
  'response',
  'from',
  'flipkart',
  '.',
  'fully',
  'regretted',
  ',',
  'i',
  'will',
  'uninstall',
  'the',
  'application',
  'and',
  'dont',
  'want',
  'to',
  'use',
  'flipkart',
  '.',
  'waste',
  'of',
  'time',
  ',',
  'there',
  'is',
  'no',
  'humanity',
  ',',
  'entire',
  'flipkart',
  'support',
  'i

In [237]:
noun_list = []
for i in token_list:
    for token in i:
        postag = nltk.pos_tag([token])[0]
        if postag[1] == 'NN':
            if postag[0] not in noun_list:
                noun_list.append(postag[0])

noun_list

['week',
 'face',
 'lock',
 'app',
 'sim',
 'iphone',
 'look',
 'apple',
 'id',
 'change',
 'cable',
 'doesnt',
 'work',
 'replacement',
 'product',
 'heating',
 'problem',
 'month',
 'phone',
 'speaker',
 'noise',
 'flipkart',
 'wrost',
 'customer',
 'service',
 'response',
 'i',
 'uninstall',
 'application',
 'dont',
 'want',
 'use',
 'waste',
 'time',
 'humanity',
 'support',
 'wrost.no',
 'team',
 '..',
 'okay',
 'delivery',
 'experience',
 'soo',
 'need',
 'chase',
 'executive',
 'multiple',
 'product.he',
 'deliver',
 'battery',
 'health',
 'usage',
 'fake',
 'showroom',
 'percentage',
 'date',
 'type',
 'software',
 'hub',
 'packaging',
 'name',
 'packaging.a',
 'scamthe',
 'device',
 'fine',
 'lot',
 'stuck',
 'phonepay',
 'google',
 'pay',
 'backup',
 'camera',
 'performance',
 'money',
 'luck',
 'charge',
 'norma',
 'issue',
 'phonesthis',
 'honest',
 'review',
 'night',
 'havent',
 'quality',
 'photography',
 'k',
 'mobile',
 'cant',
 'understand',
 'star',
 'rating',
 'retu

In [238]:
#List of unique nouns

nouns_unique=list(set(noun_list))
nouns_unique


['blank',
 't',
 'uninstall',
 'lock',
 'solve',
 'handset',
 'expirence',
 'mean',
 'filter',
 'handle',
 'regret',
 'hype',
 'option',
 'brought',
 'lack',
 'show',
 'hand',
 '☺️',
 'phone.i',
 'highlight',
 'print',
 'superb',
 'apple',
 'privacy',
 'packaging.a',
 'blood',
 'good.new',
 'cell',
 'security',
 'v',
 'offline',
 'write',
 '😙',
 'nothing',
 'fall',
 'w',
 'it.i',
 'penny',
 'everything',
 'users.please',
 'series',
 'lo',
 'launch',
 'smallcamera',
 'clarity',
 'remain',
 'goodperformance',
 'responsibility',
 'trust',
 'team',
 'play',
 'bbd.will',
 'blowing',
 'honest',
 'manager',
 'amazingtop',
 'humanity',
 'size',
 '…',
 'chipset',
 'min',
 'response',
 'delevry',
 'boring',
 'dream',
 'week',
 'sake',
 'bummer',
 'anyone',
 'fix',
 'exp',
 'temprature',
 'flipcart',
 'u',
 'impression',
 'adapter.😒',
 'reset',
 'biggbillion',
 'photo',
 'google',
 '📸',
 'try',
 '😉',
 'front',
 'super',
 'zoom',
 'andriod',
 'satisfactory',
 '🌹❤️❤️',
 '✌️',
 'schedule',
 'fingerp

In [239]:
# Removing pronouns and single alphabetical words

In [240]:
alphabets=[]
for i in string.ascii_lowercase:
    alphabets.append(i)
for i in string.ascii_uppercase:
    alphabets.append(i)

In [241]:
for i in nouns_unique:
    if i in alphabets:
        nouns_unique.remove(i)
nouns_unique

['blank',
 'uninstall',
 'lock',
 'solve',
 'handset',
 'expirence',
 'mean',
 'filter',
 'handle',
 'regret',
 'hype',
 'option',
 'brought',
 'lack',
 'show',
 'hand',
 '☺️',
 'phone.i',
 'highlight',
 'print',
 'superb',
 'apple',
 'privacy',
 'packaging.a',
 'blood',
 'good.new',
 'cell',
 'security',
 'offline',
 'write',
 '😙',
 'nothing',
 'fall',
 'it.i',
 'penny',
 'everything',
 'users.please',
 'series',
 'lo',
 'launch',
 'smallcamera',
 'clarity',
 'remain',
 'goodperformance',
 'responsibility',
 'trust',
 'team',
 'play',
 'bbd.will',
 'blowing',
 'honest',
 'manager',
 'amazingtop',
 'humanity',
 'size',
 '…',
 'chipset',
 'min',
 'response',
 'delevry',
 'boring',
 'dream',
 'week',
 'sake',
 'bummer',
 'anyone',
 'fix',
 'exp',
 'temprature',
 'flipcart',
 'impression',
 'adapter.😒',
 'reset',
 'biggbillion',
 'photo',
 'google',
 '📸',
 'try',
 '😉',
 'front',
 'super',
 'zoom',
 'andriod',
 'satisfactory',
 '🌹❤️❤️',
 '✌️',
 'schedule',
 'fingerprint',
 'ecosystem',
 'c

In [242]:
'good' in nouns_unique

False

In [243]:
#function to remove emojis

#def deEmojify(list_of_nouns):
#    final_nouns = []
#    for i in list_of_nouns:
#        review = i.encode('ascii', 'ignore').decode('ascii')
#        final_nouns.append(review)
#    return final_nouns

In [244]:
#final_nouns=deEmojify(nouns_unique)
#final_nouns

In [246]:
#pip install clean-text

In [208]:
#import clean function
#from cleantext import clean

#print text after removing the emojis from it
#nouns_unique=list(clean(nouns_unique, no_emoji=True))

In [2]:
nouns_unique

NameError: name 'nouns_unique' is not defined

In [248]:
[i for i in nouns_unique if 'good' in i]

['good.new',
 'goodperformance',
 'goodbattery',
 'gooddisplay',
 'good😊',
 'good.but',
 'androidoverpricedprosgood',
 'goodcamera',
 'goodbuy']

In [173]:
#calling the function to remove emojis

#reviews = deEmojify(nouns_unique)

In [250]:
for i in nouns_unique:
    if i==' ':
        nouns_unique.remove(i)

In [251]:
nouns_unique

['blank',
 'uninstall',
 'lock',
 'solve',
 'handset',
 'expirence',
 'mean',
 'filter',
 'handle',
 'regret',
 'hype',
 'option',
 'brought',
 'lack',
 'show',
 'hand',
 '☺️',
 'phone.i',
 'highlight',
 'print',
 'superb',
 'apple',
 'privacy',
 'packaging.a',
 'blood',
 'good.new',
 'cell',
 'security',
 'offline',
 'write',
 '😙',
 'nothing',
 'fall',
 'it.i',
 'penny',
 'everything',
 'users.please',
 'series',
 'lo',
 'launch',
 'smallcamera',
 'clarity',
 'remain',
 'goodperformance',
 'responsibility',
 'trust',
 'team',
 'play',
 'bbd.will',
 'blowing',
 'honest',
 'manager',
 'amazingtop',
 'humanity',
 'size',
 '…',
 'chipset',
 'min',
 'response',
 'delevry',
 'boring',
 'dream',
 'week',
 'sake',
 'bummer',
 'anyone',
 'fix',
 'exp',
 'temprature',
 'flipcart',
 'impression',
 'adapter.😒',
 'reset',
 'biggbillion',
 'photo',
 'google',
 '📸',
 'try',
 '😉',
 'front',
 'super',
 'zoom',
 'andriod',
 'satisfactory',
 '🌹❤️❤️',
 '✌️',
 'schedule',
 'fingerprint',
 'ecosystem',
 'c

In [252]:
'good' in nouns_unique

False

## 6. Create a review-noun binary matrix

In [253]:
#Binary Data (for association rules mining)

df_nouns={}
for noun in nouns_unique:
    binary=[]
    for review in df_reviews.Reviews:
        if noun in word_tokenize(review.lower()):
            binary.append(1)
        else:
            binary.append(0)
    df_nouns[noun]=binary


In [254]:
'good' in df_nouns

False

In [255]:
# Converting into dataframe 
binary_df=pd.DataFrame(df_nouns)
binary_df

blank  uninstall  lock  solve  handset  expirence  mean  filter  handle  \
0        0          0     1      0        0          0     0       0       0   
1        0          0     0      0        0          0     0       0       0   
2        0          0     0      0        0          0     0       0       0   
3        0          0     0      0        0          0     0       0       0   
4        0          0     0      0        0          0     0       0       0   
..     ...        ...   ...    ...      ...        ...   ...     ...     ...   
742      0          0     0      0        0          0     0       0       0   
743      0          0     0      0        0          0     0       0       0   
744      0          0     0      0        0          0     0       0       0   
745      0          0     0      0        0          0     0       0       0   
746      0          0     0      0        0          0     0       0       0   

     regret  ...  future  ....  consideration  ness  executive  👌👌  .battery  \
0         0  ...       0     0              0     0          0   0         0   
1         0  ...       0     0              0     0          0   0         0   
2         0  ...       0     0              0     0          0   0         0   
3         0  ...       0     0              0     0          0   0         0   
4         0  ...       0     0              0     0          0   0         0   
..      ...  ...     ...   ...            ...   ...        ...  ..       ...   
742       0  ...       0     0              0     0          0   0         0   
743       0  ...       0     0              0     0          0   0         0   
744       0  ...       0     0              0     0          0   0         0   
745       0  ...       0     0              0     0          0   0         0   
746       0  ...       0     0              0     0          0   0         0   

     exchnge  inferior  truecaller  
0          0         0           0  
1          0         0           0  
2          0         0           0  
3          0         0           0  
4          0         0           0  
..       ...       ...         ...  
742        0         0           0  
743        0         0           0  
744        0         0           0  
745        0         0           0  
746        0         0           0  

[747 rows x 797 columns]

In [1]:
'good' in binary_df.columns

NameError: name 'binary_df' is not defined

## 7. Extract frequent features using Apriori algorithm

In [257]:
#Association Rules Mining to extract the frequent noun phrase

In [258]:
from mlxtend.frequent_patterns import apriori, association_rules

In [261]:
# minimum support=2%

apr_df = apriori(binary_df, min_support = 0.02, use_colnames = True, max_len=2, verbose = 0)

C:\Users\HP\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [262]:
apr_df

support               itemsets
0   0.032129                (apple)
1   0.020080                  (use)
2   0.030790                (issue)
3   0.038822                 (time)
4   0.052209              (android)
5   0.028112               (mobile)
6   0.033467                (price)
7   0.020080               (backup)
8   0.029451                   (..)
9   0.025435                (money)
10  0.036145            (excellent)
11  0.065596          (performance)
12  0.066934              (battery)
13  0.208835                (phone)
14  0.020080                  (day)
15  0.065596             (flipkart)
16  0.022758                (worth)
17  0.032129              (quality)
18  0.089692               (iphone)
19  0.078983              (awesome)
20  0.049531           (experience)
21  0.132530              (product)
22  0.028112              (display)
23  0.109772               (camera)
24  0.020080                 (love)
25  0.024096                 (dont)
26  0.034806             (delivery)
27  0.024096              (heating)
28  0.020080      (iphone, android)
29  0.025435  (performance, camera)
30  0.020080       (battery, phone)
31  0.026774      (camera, battery)
32  0.020080      (flipkart, phone)
33  0.032129        (camera, phone)
34  0.020080    (flipkart, product)
35  0.021419       (iphone, camera)

## Compactness pruning

In [263]:
# Single itemsets
apr_df1=apriori(binary_df, min_support = 0.02, use_colnames = True, max_len=1, verbose = 0)
apr_df1

C:\Users\HP\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support       itemsets
0   0.032129        (apple)
1   0.020080          (use)
2   0.030790        (issue)
3   0.038822         (time)
4   0.052209      (android)
5   0.028112       (mobile)
6   0.033467        (price)
7   0.020080       (backup)
8   0.029451           (..)
9   0.025435        (money)
10  0.036145    (excellent)
11  0.065596  (performance)
12  0.066934      (battery)
13  0.208835        (phone)
14  0.020080          (day)
15  0.065596     (flipkart)
16  0.022758        (worth)
17  0.032129      (quality)
18  0.089692       (iphone)
19  0.078983      (awesome)
20  0.049531   (experience)
21  0.132530      (product)
22  0.028112      (display)
23  0.109772       (camera)
24  0.020080         (love)
25  0.024096         (dont)
26  0.034806     (delivery)
27  0.024096      (heating)

In [264]:
# Dual itemsets

apr_df2=apr_df.iloc[len(apr_df1):]
apr_df2

support               itemsets
28  0.020080      (iphone, android)
29  0.025435  (performance, camera)
30  0.020080       (battery, phone)
31  0.026774      (camera, battery)
32  0.020080      (flipkart, phone)
33  0.032129        (camera, phone)
34  0.020080    (flipkart, product)
35  0.021419       (iphone, camera)

In [265]:
# Listing out the dual itemsets

dual_features = [list(i) for i in apr_df2['itemsets']]
print(dual_features)
print(len(dual_features))

[['iphone', 'android'], ['performance', 'camera'], ['battery', 'phone'], ['camera', 'battery'], ['flipkart', 'phone'], ['camera', 'phone'], ['flipkart', 'product'], ['iphone', 'camera']]
8


In [266]:
# Maximizing the disp size

pd.set_option('display.max_rows',200)

In [267]:
# Function to do compactness pruning

compact_features=[]
def compactness_prune(w1,w2):
    
    conjunctions=['for','and','nor','but','or','yet','so','.'] # Adding full stop also here because compact words cannot be b/w 
                                                                     # two sentences
    iter_count=0

    for i in range(len(df_reviews.Reviews)):
        
        
        count=0
        token=word_tokenize(df_reviews.Reviews[i])

        if w1 in token and w2 in token:

            min_index=(min(token.index(w2),token.index(w1)))+1
            max_index=max(token.index(w2),token.index(w1))

            if not any(word in token[min_index:max_index] for word in conjunctions):
                count=len(token[min_index:max_index])

                if count<=3:
                    iter_count+=1
                    
    if iter_count>=2:
        compact_features.append([w1,w2])

                    #print(df_reviews.Reviews[i]," ----",w1,"&",w2,"are Compact\n")

In [268]:
# List of compact features which met our conditions

compact_features=[]
for features in dual_features:
    compactness_prune(str(features[0]),str(features[1]))
print(compact_features)   
print(len(compact_features))

[['iphone', 'android'], ['performance', 'camera'], ['battery', 'phone'], ['camera', 'battery'], ['camera', 'phone'], ['flipkart', 'product']]
6


In [269]:
# List of dual_features which does not met the criterion and pruned

pruned_features=[]
for i in dual_features:
     if i not in compact_features:
            pruned_features.append(i)
print(pruned_features)
print(len(pruned_features))

[['flipkart', 'phone'], ['iphone', 'camera']]
2


## Redundancy pruning

In [282]:
apr_df_redundant=apriori(binary_df, min_support = 0.02, use_colnames = True, max_len=1, verbose = 0)
apr_df_redundant

C:\Users\HP\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support       itemsets
0   0.032129        (apple)
1   0.020080          (use)
2   0.030790        (issue)
3   0.038822         (time)
4   0.052209      (android)
5   0.028112       (mobile)
6   0.033467        (price)
7   0.020080       (backup)
8   0.029451           (..)
9   0.025435        (money)
10  0.036145    (excellent)
11  0.065596  (performance)
12  0.066934      (battery)
13  0.208835        (phone)
14  0.020080          (day)
15  0.065596     (flipkart)
16  0.022758        (worth)
17  0.032129      (quality)
18  0.089692       (iphone)
19  0.078983      (awesome)
20  0.049531   (experience)
21  0.132530      (product)
22  0.028112      (display)
23  0.109772       (camera)
24  0.020080         (love)
25  0.024096         (dont)
26  0.034806     (delivery)
27  0.024096      (heating)

In [283]:
# Function to find redundant features
rev=[]
redundant_features=[]

def redundance_prune(w1,w2):
    
    #conjunctions=['for','and','nor','but','or','yet','so']
    #iter_count=0

    for i in range(len(df_reviews.Reviews)):
        
        
        count=0
        token=word_tokenize(df_reviews.Reviews[i])

        if w1 in token and w2 in token:

            min_index=(min(token.index(w2),token.index(w1)))
            max_index=max(token.index(w2),token.index(w1))

            #if not any(word in token[min_index:max_index] for word in conjunctions):
            count=len(token[min_index:max_index])

            if count==1:
                redundant_features.append([w1,w2])
                rev.append(df_reviews.Reviews[i])
                #iter_count+=1
                    
    #if iter_count>0:
        

                    #print(df_reviews.Reviews[i]," ----",w1,"&",w2,"are Compact\n")

In [284]:
# List of reduntant features with corresponding reviews


rev=[]        # reviews corresponding to feature
redundant_features=[]

for features in dual_features:
    redundance_prune(str(features[0]),str(features[1]))

for i in range(len(rev)):
    print(redundant_features[i])   
    print(rev[i])

    

['performance', 'camera']
poor camera performance
['performance', 'camera']
best in term of camera performance
['performance', 'camera']
i purchase this iphone  a month ago. no need to tell about performance camera and design it is top notch. but the main highlight and the brilliant improvement in its battery.
['camera', 'battery']
i like battery camera
['camera', 'phone']
premium phone camera 😍
['camera', 'phone']
thank flipkart for very best price and very fast delivery. i phone  camera quality is good.
['flipkart', 'product']
with in a month the phone speaker is giving some annoying noise, really disappointed with the flipkart product


In [285]:
# List of reduntant features
redundant_features

[['performance', 'camera'],
 ['performance', 'camera'],
 ['performance', 'camera'],
 ['camera', 'battery'],
 ['camera', 'phone'],
 ['camera', 'phone'],
 ['flipkart', 'product']]

In [286]:
# Function to calculate the count of reduntant features
features_freq={}
def reduntant_features_count(item):
    count=0
    for i in range(len(redundant_features)):
        if item in redundant_features[i][0] or item in redundant_features[i][1]:
            count+=1
    features_freq[item]=count
    return features_freq

In [287]:
frozenset_conv=[tuple(i) for i in apr_df_redundant.itemsets]
frozenset_conv

[('apple',),
 ('use',),
 ('issue',),
 ('time',),
 ('android',),
 ('mobile',),
 ('price',),
 ('backup',),
 ('..',),
 ('money',),
 ('excellent',),
 ('performance',),
 ('battery',),
 ('phone',),
 ('day',),
 ('flipkart',),
 ('worth',),
 ('quality',),
 ('iphone',),
 ('awesome',),
 ('experience',),
 ('product',),
 ('display',),
 ('camera',),
 ('love',),
 ('dont',),
 ('delivery',),
 ('heating',)]

In [288]:
# We'll find redundancy for these candidates 

overall_features=[]
for i in frozenset_conv:
    overall_features.append(i[0])
overall_features

['apple',
 'use',
 'issue',
 'time',
 'android',
 'mobile',
 'price',
 'backup',
 '..',
 'money',
 'excellent',
 'performance',
 'battery',
 'phone',
 'day',
 'flipkart',
 'worth',
 'quality',
 'iphone',
 'awesome',
 'experience',
 'product',
 'display',
 'camera',
 'love',
 'dont',
 'delivery',
 'heating']

In [289]:
nltk.pos_tag(['buy'])

[('buy', 'VB')]

In [290]:
# Overall total frequency of the below selected features

overall_freq={}
for feature in overall_features:
    count=0
    for line in df_reviews.Reviews:
        tag = nltk.pos_tag(word_tokenize(line))
        for i in range(len(tag)):
            if tag[i][0]==feature:
                count+=1
    overall_freq[feature]=count
print(overall_freq)

{'apple': 27, 'use': 18, 'issue': 24, 'time': 29, 'android': 53, 'mobile': 22, 'price': 27, 'backup': 16, '..': 30, 'money': 19, 'excellent': 28, 'performance': 51, 'battery': 53, 'phone': 183, 'day': 16, 'flipkart': 63, 'worth': 17, 'quality': 28, 'iphone': 80, 'awesome': 63, 'experience': 44, 'product': 112, 'display': 22, 'camera': 91, 'love': 15, 'dont': 18, 'delivery': 30, 'heating': 19}


In [291]:
# Redundancy pair features frequency

features_freq={}
for i in overall_features:
    reduntant_features_count(i)
features_freq 

# product awesome, product flipkart, good product etc..,

{'apple': 0,
 'use': 0,
 'issue': 0,
 'time': 0,
 'android': 0,
 'mobile': 0,
 'price': 0,
 'backup': 0,
 '..': 0,
 'money': 0,
 'excellent': 0,
 'performance': 3,
 'battery': 1,
 'phone': 2,
 'day': 0,
 'flipkart': 1,
 'worth': 0,
 'quality': 0,
 'iphone': 0,
 'awesome': 0,
 'experience': 0,
 'product': 1,
 'display': 0,
 'camera': 6,
 'love': 0,
 'dont': 0,
 'delivery': 0,
 'heating': 0}

In [299]:
# Redundancy pruning - p-support 

final_features={}
for feature in overall_features:
    count=0
    for line in df_reviews.Reviews:
        tag = nltk.pos_tag(word_tokenize(line))
        for i in range(len(tag)):
            if tag[i][0]==feature:
                count+=1
    final_features[feature]=count-features_freq[feature]      # product - [product awesome, product flipkart, good product,..]
print(final_features)

{'apple': 27, 'use': 18, 'issue': 24, 'time': 29, 'android': 53, 'mobile': 22, 'price': 27, 'backup': 16, '..': 30, 'money': 19, 'excellent': 28, 'performance': 48, 'battery': 52, 'phone': 181, 'day': 16, 'flipkart': 62, 'worth': 17, 'quality': 28, 'iphone': 80, 'awesome': 63, 'experience': 44, 'product': 111, 'display': 22, 'camera': 85, 'love': 15, 'dont': 18, 'delivery': 30, 'heating': 19}


In [300]:
final_features.keys()

dict_keys(['apple', 'use', 'issue', 'time', 'android', 'mobile', 'price', 'backup', '..', 'money', 'excellent', 'performance', 'battery', 'phone', 'day', 'flipkart', 'worth', 'quality', 'iphone', 'awesome', 'experience', 'product', 'display', 'camera', 'love', 'dont', 'delivery', 'heating'])

In [301]:
removable_items=['apple','use','issue','time','android','mobile','..','excellent','phone','day','awesome','flipkart','worth','iphone','experience','product','delivery','love','dont']

In [302]:
[final_features.pop(keys) for keys in removable_items]

[27, 18, 24, 29, 53, 22, 30, 28, 181, 16, 63, 62, 17, 80, 44, 111, 30, 15, 18]

In [303]:
final_features

{'price': 27,
 'backup': 16,
 'money': 19,
 'performance': 48,
 'battery': 52,
 'quality': 28,
 'display': 22,
 'camera': 85,
 'heating': 19}

###  Comment on the level of automation achieved using this framework

* <b> From the compactness pruning, we removed the candidate features containing two words that are likely to be meaningless

* <b> From the p-support of every single feature, we removed the candidate feature containing one word that is likely to be meaningless

### Discuss the major advantages of this process

### Three key benefits of feature selection are:
    
* <b> Fewer redundant data means fewer chances of making decisions based on noise.
* <b> Less misleading data means better modeling accuracy.
* <b> Reduces Training Time. Less data means quicker algorithms

### Discuss the major limitations of the process

* <b> Automization is not possible in few processes, more manual work is required 
* <b> Time consuming process while creating binary dataframes for large amount of reviews 
* <b> Space and model complexity is also high 
* <b> Difficult to assess the results of algorithm while handling large corpus of data

In [304]:
### Hence we would give 4.5/5 rating for this framework 

In [305]:
final_features

{'price': 27,
 'backup': 16,
 'money': 19,
 'performance': 48,
 'battery': 52,
 'quality': 28,
 'display': 22,
 'camera': 85,
 'heating': 19}

In [306]:
list(final_features.keys())

['price',
 'backup',
 'money',
 'performance',
 'battery',
 'quality',
 'display',
 'camera',
 'heating']

In [307]:
#for sentence in df_reviews.Reviews:
    #if 'product' in sentence:
     #   for j in range(len(nltk.pos_tag(word_tokenize(sentence.lower())))):
      #      if [nltk.pos_tag([i]) for i in word_tokenize(sentence.lower())][j][0][1]==('JJ' or 'JJR' or 'JJS'):
       #         print([nltk.pos_tag([i]) for i in word_tokenize(sentence.lower())][j][0][0])
            
        

## Opinion Extraction

In [308]:
finalized_features=list(final_features.keys())
finalized_features

['price',
 'backup',
 'money',
 'performance',
 'battery',
 'quality',
 'display',
 'camera',
 'heating']

In [339]:
opinion_list = []
print("Features:")
for f in finalized_features:
    sent_ind = -1
    print(f)
 
    for sentence in token_list:
        opinion1 = ''
        opinion2 = ''
        sent_ind +=1
       
        if f in sentence:
            f_ind = sentence.index(f)


            #print(f_ind)
            mini1 = 10e100
            mini2 = 10e100
            
            for i in range(f_ind, -1, -1):
                if 'JJ' in nltk.pos_tag([sentence[i]])[0][1]:
                    mini1 = i
                    break
                    #print(min1)
            for i in range(f_ind, len(sentence)):
                if 'JJ' in nltk.pos_tag([sentence[i]])[0][1]:
                    mini2 = i
                    break
                    #print(min1)
            #print(mini1, mini2)
            d1 = abs(f_ind - mini1)
            d2 = abs(mini2 - f_ind)
            if d1 < d2:
                mini = mini1
            else:
                mini = mini2
            #print(d1, d2)
            #print(mini, "ppppp")
            if mini != 10e100:
                #print(sentence[mini])                   ##Nearest adjective
               #print(mini, len(sentence)-1)
                if mini != 0:
                    if 'RB' in nltk.pos_tag([sentence[mini-1]])[0][1]:
                        opinion1 = sentence[mini-1] +' '+ sentence[mini]
                    elif 'JJ' in nltk.pos_tag([sentence[mini-1]])[0][1]:
                        opinion1 = sentence[mini-1] +' '+sentence[mini]
                if mini != len(sentence)-1:
                    if 'JJ' in nltk.pos_tag([sentence[mini+1]])[0][1]:
                        opinion1 = sentence[mini] + ' '+ sentence[mini+1]
                if opinion1 == '':
                    opinion1 = sentence[mini]
                m1 = mini    
                #print(opinion1, sent_ind)
    # ----------------------------------------------------------------------------------------------------------
            mini1 = 10e100
            mini2 = 10e100
            lst = ['am', 'be', 'is','was','are','were','has', 'have', 'had','been', 'say']
            for i in range(f_ind, -1, -1):
                if sentence[i] not in lst:
                    if 'VB' in nltk.pos_tag([sentence[i]])[0][1]:
                        #print('s')
                        mini1 = i
                        break
                        #print(min1)
            for i in range(f_ind, len(sentence)):
                if sentence[i] not in lst:
                    if 'VB' in nltk.pos_tag([sentence[i]])[0][1]:
                        #print('s')
                        mini2 = i
                        break
                        #print(min1)
            #print(mini1, mini2)
            d1 = abs(f_ind - mini1)
            d2 = abs(mini2 - f_ind)
            if d1 < d2:
                mini = mini1
            else:
                mini = mini2
            #print(d1, d2)
            #print(mini, "ppppp")
            if mini != 10e100:
                #print(sentence[mini])                   ##Nearest adverb
               #print(mini, len(sentence)-1)

                if mini != 0:
                    if 'RB' in nltk.pos_tag([sentence[mini-1]])[0][1]:
                        opinion2 = sentence[mini-1] +' '+ sentence[mini]
                if mini != len(sentence)-1:
                    if 'JJ' in nltk.pos_tag([sentence[mini+1]])[0][1]:
                        opinion2 = sentence[mini] + ' '+ sentence[mini+1]
                    elif 'RB' in nltk.pos_tag([sentence[mini+1]])[0][1]:
                        opinion2 = sentence[mini] + ' '+ sentence[mini+1]
                if opinion2 == '':
                    opinion2 = sentence[mini]
                m2 = mini  
                #print(opinion2, sent_ind)



                dis1 = abs(f_ind - m1)
                dis2 = abs(f_ind - m2)
                if dis1 > dis2:
                    opinion = opinion2
                elif dis1 < dis2:
                    opinion = opinion1
                else:
                    import numpy as np
                    opinion = np.random.choice([opinion1, opinion2])


                if opinion != '' and opinion[0].isnumeric() != True:
                    #print(opinion, sent_ind)
                    if opinion not in opinion_list:
                        opinion_list.append(opinion)
opinion_list

Features:
price
backup
money
performance
battery
quality
display
camera
heating


['considering',
 'facing',
 'too much',
 'high',
 'not expected',
 'expect good',
 'good',
 'very small',
 'expected',
 'promised',
 'reasonable',
 'amazing',
 'available',
 'very good',
 'delivering',
 'worstest',
 'also good',
 'so good',
 'less compared',
 'nice',
 'also improved',
 'happy',
 'complete',
 'earned',
 'compared',
 'returned',
 'buy',
 'many',
 'go',
 'very bad',
 'charging',
 'somewhat decreased',
 'most',
 'waiting',
 'overall good',
 'loved',
 'used',
 'too good',
 'very satisfying',
 'full',
 'very nice',
 'difficult',
 'simply unmatched',
 'great',
 'totally satisfied',
 'really fantastic',
 'other',
 'gets down',
 'draining',
 'felling',
 'last',
 'desing',
 'better optimized',
 'fabulous',
 'bad',
 'recommended',
 'started',
 'little',
 'enhanced',
 'best',
 'added',
 'fully enjoying',
 'get',
 'poor',
 'not satisfied',
 'using',
 'when other',
 'low',
 'know',
 'very slow',
 'hanging',
 'not working',
 'handling',
 'worst',
 'take',
 'exciting',
 'cinematic',
 

In [312]:
len(opinion_list)

97

## Cleaning opinions using enchant dictionary

In [366]:
! pip install PyEnchant

In [313]:
import enchant

In [314]:
d=enchant.Dict("en_US")

In [315]:
d.check('worstest')

False

In [316]:
single_words=[i for i in opinion_list if " " not in i]
len(single_words)

67

In [317]:
double_words=[i.split() for i in opinion_list if " " in i]
len(double_words)

30

In [318]:
split_check1=[i for i in single_words if d.check(i)==True]
len(split_check1)

64

In [319]:
split_check2=[]
for i in range(len(double_words)):
    if d.check(double_words[i][0]) and d.check(double_words[i][1]):
        split_check2.append(double_words[i])
len(split_check2)

30

In [320]:
rejoin=[" ".join(i) for i in double_words]
rejoin

['too much',
 'not expected',
 'not available',
 'very small',
 'very good',
 'also good',
 'so good',
 'less compared',
 'also improved',
 'very bad',
 'somewhat decreased',
 'overall good',
 'too good',
 'very satisfying',
 'very nice',
 'simply unmatched',
 'totally satisfied',
 'really fantastic',
 'gets down',
 'better optimized',
 'fully enjoying',
 'not satisfied',
 'when other',
 'very slow',
 'not working',
 'very outdated',
 'also does',
 'really amazing',
 'very happy',
 'charging ideally']

In [321]:
opinion_final=split_check1+rejoin
len(opinion_final)

94

## Opinion orientation identification

# Sentiment analyser

In [574]:
# Getting the sentiment scores for five reviews
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

general_opinion={}
pos_opinion={}
neg_opinion={}
for opinion in opinion_final:
    senti=SentimentIntensityAnalyzer() ### Sentiment Analysis using VADER
    #print(opinion)
    score=senti.polarity_scores(opinion).get('compound')
    if score!=0:
        general_opinion[opinion]=score
    if score>0:                        # Here considering 0 and above scores as positive
        pos_opinion[opinion]=score
    elif score<0:
        neg_opinion[opinion]=score
    else:
        pass  # Neutral

In [575]:
df_positive_opinion=pd.DataFrame(pos_opinion.items(),columns=['Positive Opinion','Polarity scores'])
df_positive_opinion

Positive Opinion  Polarity scores
0                good           0.4404
1            promised           0.3612
2             amazing           0.5859
3                nice           0.4215
4               happy           0.5719
5               loved           0.5994
6               great           0.6249
7            fabulous           0.5267
8         recommended           0.2023
9                best           0.6369
10           exciting           0.4939
11           stunning           0.3818
12          very good           0.4927
13          also good           0.4404
14            so good           0.4927
15      also improved           0.4767
16       overall good           0.4404
17           too good           0.4404
18    very satisfying           0.5095
19          very nice           0.4754
20  totally satisfied           0.4754
21   really fantastic           0.5984
22   better optimized           0.7096
23     fully enjoying           0.5709
24     really amazing           0.6240
25         very happy           0.6115
26   charging ideally           0.4215

In [576]:
df_negative_opinion=pd.DataFrame(neg_opinion.items(),columns=['Negative Opinion','Polarity scores'])
df_negative_opinion

Negative Opinion  Polarity scores
0          difficult          -0.3612
1                bad          -0.5423
2               poor          -0.4767
3                low          -0.2732
4              worst          -0.6249
5            refused          -0.2960
6           annoying          -0.4019
7               hard          -0.1027
8       disappointed          -0.4767
9            lagging          -0.2732
10          very bad          -0.5849
11  simply unmatched          -0.0772
12     not satisfied          -0.3252

In [577]:
df_overall_opinion=pd.DataFrame(general_opinion.items(),columns=['Overall Opinions','Polarity scores'])
df_overall_opinion

Overall Opinions  Polarity scores
0                good           0.4404
1            promised           0.3612
2             amazing           0.5859
3                nice           0.4215
4               happy           0.5719
5               loved           0.5994
6           difficult          -0.3612
7               great           0.6249
8            fabulous           0.5267
9                 bad          -0.5423
10        recommended           0.2023
11               best           0.6369
12               poor          -0.4767
13                low          -0.2732
14              worst          -0.6249
15           exciting           0.4939
16            refused          -0.2960
17           annoying          -0.4019
18               hard          -0.1027
19       disappointed          -0.4767
20           stunning           0.3818
21            lagging          -0.2732
22          very good           0.4927
23          also good           0.4404
24            so good           0.4927
25      also improved           0.4767
26           very bad          -0.5849
27       overall good           0.4404
28           too good           0.4404
29    very satisfying           0.5095
30          very nice           0.4754
31   simply unmatched          -0.0772
32  totally satisfied           0.4754
33   really fantastic           0.5984
34   better optimized           0.7096
35     fully enjoying           0.5709
36      not satisfied          -0.3252
37     really amazing           0.6240
38         very happy           0.6115
39   charging ideally           0.4215

## Summary generation

In [578]:
total_positive_polarity=np.sum(df_positive_opinion['Polarity scores'])
total_positive_polarity

13.6263

In [579]:
total_negative_polarity=-1*np.sum(df_negative_opinion['Polarity scores'])
total_negative_polarity

4.8161000000000005

In [580]:
postive_percent=100*(total_positive_polarity/(total_positive_polarity+total_negative_polarity))
postive_percent

73.88571986292457

In [581]:
negative_percent=100*(total_negative_polarity/(total_positive_polarity+total_negative_polarity))
negative_percent

26.11428013707544

## Infrequent feature mining

In [546]:
infreq=[]
for sent in df_reviews.Reviews:
    for f in finalized_features:
        if f not in sent:
            for o in opinion_final:
                if o in word_tokenize(sent.lower()):
                    #print(sent)
                    #print(o)
                    sentence=word_tokenize(sent.lower())
    
                    if len(word_tokenize(o))==2:
                        op=word_tokenize(o)[1]
                    else:
                        op=o
                    
                    opinion_index=sentence.index(op)

                    left_noun=[]
                    if opinion_index!=0:
                        left_noun=sentence[opinion_index-1::-1]

                    right_noun=[]
                    if opinion_index==0 or opinion_index!=len(sentence)-1:
                        right_noun=sentence[opinion_index+1::1]

                    #print(left_noun)
                    #print(right_noun)
                    noun1,noun2=[],[]
                    if len(left_noun)!=0:

                        leftn=[i for i in left_noun]

                        for i in range(len(leftn)):
                            k=nltk.pos_tag([str(leftn[i])])
                            if k[0][1]=='NN':
                                noun1=leftn[i]

                        #noun1=[i for i in nltk.pos_tag(leftn) if 'NN' in i ]
                        #print('yes')
                        #print([i for i in nltk.pos_tag(left_noun) if 'NN' in i ])
                        #if len(noun1)!=0:
                         #   noun1=noun1[0][0]
                      #      print(noun1)


                    if len(right_noun)!=0:  

                        rightn=[[i] for i in right_noun]
                        for i in range(len(rightn)):
                            k=nltk.pos_tag([str(rightn[i])])
                            if k[0][1]=='NN':
                                noun2=rightn[i]

                    if noun1 and noun2 in sentence:
                        dist1=abs(opinion_index-sentence.index(noun1))
                        dist2=abs(opinion_index-sentence.index(noun2))
                        if dist1<dist2:
                            infreq.append(noun1)
                        elif dist1>dist2:
                            infreq.append(noun2)
                        else:
                            import numpy as np
                            infreq.append(np.random.choice([noun1,noun2]))
                    elif noun1 in sentence:
                        infreq.append(noun1)
                    elif noun2 in sentence:
                        infreq.append(noun2)
print(set(infreq))

{'expirence', 'hype', 'price', 'luck', 'month', 'switch', 'daycamera', 'gamingpubg', 'superb', 'please', 'apple', 'packaging', 'prosbrighter', 'ot', 'wao', 'speaker', 'wrost', 'nothing', 'thank', 'review', 'reality', 'everything', 'fine', 'series', 'camera', 'backup', 'screen', 'system', 'body', 'waste', 'dream', 'picture', 'i', 'week', 'impression', 'front', 'super', 'zoom', 'move', 'genuine', 'st', 'wireless', 'flipkart', 'thing', 'brand', 'im', 'need', 'excellent', 'beautiful', 'dont', 'box', 'perfect', 'mid', 'journey', 'beware', 'year', 'wow', 'colour', 'performance', 'delivery', 'soo', 'impactful', 'battery', 'quality', 'iphone', 'omg', 'power', 'damn', 'phone', 'awesome', 'heat', 'cemera', '’', 'kind', 'sexy', 'heating', 'time', 'place', 'osm', 'experience', 'bbd', 'verry', 'smooth', 'brilliant', 'look', 'buttery', 'android', 'brightness', 'purchase', 'bought', 'product', 'display', 'k', 'mode', 'mobile'}


In [549]:
infreq_features=list(set(infreq))
len(infreq_features)

95

In [550]:
infreq_features

['expirence',
 'hype',
 'price',
 'luck',
 'month',
 'switch',
 'daycamera',
 'gamingpubg',
 'superb',
 'please',
 'apple',
 'packaging',
 'prosbrighter',
 'ot',
 'wao',
 'speaker',
 'wrost',
 'nothing',
 'thank',
 'review',
 'reality',
 'everything',
 'fine',
 'series',
 'camera',
 'backup',
 'screen',
 'system',
 'body',
 'waste',
 'dream',
 'picture',
 'i',
 'week',
 'impression',
 'front',
 'super',
 'zoom',
 'move',
 'genuine',
 'st',
 'wireless',
 'flipkart',
 'thing',
 'brand',
 'im',
 'need',
 'excellent',
 'beautiful',
 'dont',
 'box',
 'perfect',
 'mid',
 'journey',
 'beware',
 'year',
 'wow',
 'colour',
 'performance',
 'delivery',
 'soo',
 'impactful',
 'battery',
 'quality',
 'iphone',
 'omg',
 'power',
 'damn',
 'phone',
 'awesome',
 'heat',
 'cemera',
 '’',
 'kind',
 'sexy',
 'heating',
 'time',
 'place',
 'osm',
 'experience',
 'bbd',
 'verry',
 'smooth',
 'brilliant',
 'look',
 'buttery',
 'android',
 'brightness',
 'purchase',
 'bought',
 'product',
 'display',
 'k',


# Sentiment analysis for opinions about infrequent features

In [582]:
# Getting the sentiment scores for five reviews
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

general_opinion2={}
pos_opinion2={}
neg_opinion2={}
for opinion in infreq_features:
    senti=SentimentIntensityAnalyzer() ### Sentiment Analysis using VADER
    #print(opinion)
    score=senti.polarity_scores(opinion).get('compound')
    if score!=0:
        general_opinion2[opinion]=score
    if score>0:                           # Here considering above 0 scores as positive
        pos_opinion2[opinion]=score
    elif score<0:
        neg_opinion2[opinion]=score
    else:
        pass  # Neutral

In [583]:
df_positive_opinion2=pd.DataFrame(pos_opinion2.items(),columns=['Positive Opinion','Polarity scores'])
df_positive_opinion2

Positive Opinion  Polarity scores
0              luck           0.4588
1            superb           0.6249
2            please           0.3182
3             thank           0.3612
4              fine           0.2023
5             dream           0.2500
6        impression           0.2263
7             super           0.5994
8         excellent           0.5719
9         beautiful           0.5994
10          perfect           0.5719
11              wow           0.5859
12          awesome           0.6249
13             kind           0.5267
14             sexy           0.5267
15        brilliant           0.5859
16       brightness           0.3818

In [585]:
df_negative_opinion2=pd.DataFrame(neg_opinion2.items(),columns=['Negative Opinion','Polarity scores'])
df_negative_opinion2

Negative Opinion  Polarity scores
0            waste          -0.4215
1             damn          -0.4019

In [586]:
df_overall_opinion2=pd.DataFrame(general_opinion2.items(),columns=['Overall Opinions','Polarity scores'])
df_overall_opinion2

Overall Opinions  Polarity scores
0              luck           0.4588
1            superb           0.6249
2            please           0.3182
3             thank           0.3612
4              fine           0.2023
5             waste          -0.4215
6             dream           0.2500
7        impression           0.2263
8             super           0.5994
9         excellent           0.5719
10        beautiful           0.5994
11          perfect           0.5719
12              wow           0.5859
13             damn          -0.4019
14          awesome           0.6249
15             kind           0.5267
16             sexy           0.5267
17        brilliant           0.5859
18       brightness           0.3818

## Summary generation

In [587]:
total_positive_polarity2=np.sum(df_positive_opinion2['Polarity scores'])
total_positive_polarity2

8.0162

In [588]:
total_negative_polarity2=-1*np.sum(df_negative_opinion2['Polarity scores'])
total_negative_polarity2

0.8233999999999999

In [589]:
postive_percent2=100*(total_positive_polarity2/(total_positive_polarity2+total_negative_polarity2))
postive_percent2

90.68509887325219

In [590]:
negative_percent2=100*(total_negative_polarity2/(total_positive_polarity2+total_negative_polarity2))
negative_percent2

9.314901126747817

## Combining both frequent and infrequent features opinion

In [600]:
df_positive_combined=df_positive_opinion.append(df_positive_opinion2,ignore_index=True)
df_positive_combined

Positive Opinion  Polarity scores
0              good           0.4404
1          promised           0.3612
2           amazing           0.5859
3              nice           0.4215
4             happy           0.5719
..              ...              ...
39          awesome           0.6249
40             kind           0.5267
41             sexy           0.5267
42        brilliant           0.5859
43       brightness           0.3818

[44 rows x 2 columns]

In [599]:
# We'll set max rows to just 10 for visualization complexity

pd.set_option('display.max_rows',10)

In [601]:
df_negative_combined=df_negative_opinion.append(df_negative_opinion2,ignore_index=True)
df_negative_combined

Negative Opinion  Polarity scores
0          difficult          -0.3612
1                bad          -0.5423
2               poor          -0.4767
3                low          -0.2732
4              worst          -0.6249
..               ...              ...
10          very bad          -0.5849
11  simply unmatched          -0.0772
12     not satisfied          -0.3252
13             waste          -0.4215
14              damn          -0.4019

[15 rows x 2 columns]

In [602]:
df_combined_opinions=df_overall_opinion.append(df_overall_opinion2,ignore_index=True)
df_combined_opinions

Overall Opinions  Polarity scores
0              good           0.4404
1          promised           0.3612
2           amazing           0.5859
3              nice           0.4215
4             happy           0.5719
..              ...              ...
54          awesome           0.6249
55             kind           0.5267
56             sexy           0.5267
57        brilliant           0.5859
58       brightness           0.3818

[59 rows x 2 columns]

## Summary generation

In [603]:
overall_positive_polarity=np.sum(df_positive_combined['Polarity scores'])
overall_positive_polarity

21.642499999999995

In [604]:
overall_negative_polarity=-1*np.sum(df_negative_combined['Polarity scores'])
overall_negative_polarity

5.639500000000001

In [605]:
overall_postive_percent=100*(overall_positive_polarity/(overall_positive_polarity+overall_negative_polarity))
overall_postive_percent

79.32886152041638

In [607]:
overall_negative_percent=100*(overall_negative_polarity/(overall_positive_polarity+overall_negative_polarity))
overall_negative_percent

20.671138479583615

## Conclusion about iphone13 reviews sentiment analysis

* <b> As a result of the sentiment analysis on sampled 747 reviews on iphone13, the percentage of positive and negative reviews were predicted. 
* <b> The mobile has got around 80% positive reviews.
* <b> Around 20% negative reviews were found for this product.
* <b> I would suggest iphone13 as a go to product for the new buyer whose budget is around 60-65k.
